File: python/echoClient/eventViewer.py
#!/usr/bin/env python
import sys
sys.path.append("../../dataAcq/buffer/python")
import numpy
import FieldTrip
import time

def pythonclient(hostname='localhost',port=1972,timeout=5000):
        		
    ftc = FieldTrip.Client()		
    # Wait until the buffer connects correctly and returns a valid header
    hdr = None;
    while hdr is None :
        print 'Trying to connect to buffer on %s:%i ...'%(hostname,port)
        try:
            ftc.connect(hostname, port)
            print '\nConnected - trying to read header...'
            hdr = ftc.getHeader()
	except IOError:
            pass
	
        if hdr is None:
            print 'Invalid Header... waiting'
            time.sleep(1)
        else:
            print hdr
            print hdr.labels

    # Now do the echo server
    nEvents=hdr.nEvents;
    endExpt=None;
    while endExpt is None :
        (curSamp,curEvents)=ftc.wait(-1,nEvents,timeout) # Block until there are new events to process
        if curEvents>nEvents :
            evts=ftc.getEvents([nEvents,curEvents-1]) 
            nEvents=curEvents # update record of which events we've seen
            for evt in evts:
                print evt
        else:
            print "Wait timeout, waiting"	
    ftc.disconnect()


if __name__ == "__main__":
    hostname='localhost'
    port=1972
    timeout=5000    
    if len(sys.argv)>1: # called with options, i.e. commandline
        hostname = sys.argv[1]
	if len(sys.argv)>2:
            try:
                port = int(sys.argv[2])
            except:
                print 'Error: second argument (%s) must be a valid (=integer) port number'%sys.argv[2]
                sys.exit(1)
    pythonclient(hostname,port);

File: java/echoClient/EventViewer.java
import java.io.*;
import java.nio.*;
import nl.fcdonders.fieldtrip.bufferclient.*;

class EventViewer {
	 protected static final String TAG = EventViewer.class.getSimpleName();


	 public static void main(String[] args) throws IOException,InterruptedException {
		String hostname = "localhost";
		int port = 1972;
		int timeout = 500;
	
		if (args.length>=1) {
			hostname = args[0];
			int sep = hostname.indexOf(':');
			if ( sep>0 ) {
				 port=Integer.parseInt(hostname.substring(sep+1,hostname.length()));
				 hostname=hostname.substring(0,sep);
			}			
		}
		System.out.println("Host: "+hostname+":"+port);		

		if (args.length>=3) {
			try {
				port = Integer.parseInt(args[2]);
			}
			catch (NumberFormatException e) {
				timeout = 500;
			}
		}
		System.out.println(TAG+" timeout_ms: " + timeout);
		
		BufferClientClock C = new BufferClientClock();

		Header hdr=null;
		while( hdr==null ) {
			 try {
				  System.out.println(TAG+" Connecting to "+hostname+":"+port);
				  if ( !C.isConnected() ) {
						C.connect(hostname, port);
				  }
				  //C.setAutoReconnect(true);
				  if ( C.isConnected() ) { hdr = C.getHeader(); }
			 } catch (IOException e) {
				  hdr=null;
			 }
			 if ( hdr==null ){
 				  System.out.println(TAG+" Invalid Header... waiting");
				  Thread.sleep(1000);
			 }
		}
				  //float[][] data = C.getFloatData(0,hdr.nSamples-1);				
				  System.out.println("#channels....: "+hdr.nChans);
				  System.out.println("#samples.....: "+hdr.nSamples);
				  System.out.println("#events......: "+hdr.nEvents);
				  System.out.println("Sampling Freq: "+hdr.fSample);
				  System.out.println("data type....: "+hdr.dataType);
				  for (int n=0;n<hdr.nChans;n++) {
						if (hdr.labels[n] != null) {
							 System.out.println("Ch. " + n + ": " + hdr.labels[n]);
						}
				  }
						
		// Now do the event viewer
		int nEvents=hdr.nEvents;
		int nevt=0;
		int printCount=0;
		boolean endExpt=false;
		while ( !endExpt ) {
			 SamplesEventsCount sec = C.waitForEvents(nEvents,timeout); // Block until there are new events
			 if ( sec.nEvents > nEvents ){
				  // get the new events
				  BufferEvent[] evs = C.getEvents(nEvents,sec.nEvents-1);
				  nEvents=sec.nEvents;// update record of which events we've seen
				  // filter for ones we want
				  for ( int ei=0; ei<evs.length; ei++){
						BufferEvent evt=evs[ei];
						System.out.println(nevt + ") " + evt); // Print the event to the console
						nevt++;
				  }
			 } else { // timed out without new events
				  printCount+=timeout;
				  if (printCount>5000){// only print timeout message every 5s
						System.out.println(sec.nSamples + "," + sec.nEvents + " (samp,evt) Press <ENTER> to send event.");
						printCount=0;
				  }
			 }
			 if ( System.in.available()>0 ) { // keys to read
				  java.util.Scanner keyboard = new java.util.Scanner(System.in);
				  keyboard.nextLine();
				  // read the initial enter.
				  System.out.print("Enter event type:");
				  String type = keyboard.nextLine();
				  System.out.print("Enter event value:");
				  String value = keyboard.nextLine();
				  C.putEvent(new BufferEvent(type,value,-1)); 
			 }
		}
		C.disconnect();
	}
}
